src/backend/distributed/commands/multi_copy.c

```diff
@@ -2311,47 +2397,67 @@ CitusSendTupleToPlacements(TupleTableSlot *slot, CitusCopyDestReceiver *copyDest
 	/* connections hash is kept in memory context */
 	MemoryContextSwitchTo(copyDest->memoryContext);
 
-	/* get existing connections to the shard placements, if any */
-	shardConnections = GetShardHashConnections(shardConnectionHash, shardId,
-											   &shardConnectionsFound);
-	if (!shardConnectionsFound)
+	shardState = GetShardState(copyDest->shardStateHash,
+							   copyDest->connectionStateHash, shardId);
+	foreach(placementStateCell, shardState->placementStateList)
 	{
-		/*
-		 * Keep track of multi shard accesses before opening connection
-		 * the second shard.
-		 */
-		if (!copyDest->multiShardCopy && hash_get_num_entries(shardConnectionHash) == 2)
-		{
-			Oid relationId = copyDest->distributedRelationId;
+		CopyPlacementState *placementState = lfirst(placementStateCell);
+		CopyConnectionState *connectionState = placementState->connectionState;
+		bool switchOver = false;
 
-			/* mark as multi shard to skip doing the same thing over and over */
-			copyDest->multiShardCopy = true;
+		if (placementState != connectionState->activePlacementState)
+		{
+			if (placementState->data->len > COPY_SWITCH_OVER_THRESHOLD ||
+				connectionState->activePlacementState == NULL)
+			{
+				switchOver = true;
+			}
+			else
+			{
+				/* buffer data */
+				StringInfo copyBuffer;
+				copyBuffer = copyOutState->fe_msgbuf;
+				resetStringInfo(copyBuffer);
+				AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
+								  copyOutState, columnOutputFunctions,
+								  columnCoercionPaths);
+				appendBinaryStringInfo(placementState->data, copyBuffer->data,
+									   copyBuffer->len);
+			}
+		}
 
-			/* error out of conflicting COPY */
-			CheckConflictingParallelCopyAccesses(relationId);
+		if (switchOver)
+		{
+			if (connectionState->activePlacementState != NULL)
+			{
+				EndPlacementStateCopyCommand(connectionState->activePlacementState,
+											 copyOutState);
+				dlist_push_head(&connectionState->bufferedPlacementList,
+								&connectionState->activePlacementState->
+								bufferedPlacementNode);
+			}
+			StartPlacementStateCopyCommand(placementState, copyStatement,
+										   copyOutState);
```

Comment:
`connectionState->activePlacementState = placementState` is part of the switch-over operation, but `StartPlacementStateCopyCommand` doesn't know anything about switch-over. It just starts a binary/non-binary COPY command on the connection. I think we can pass a connection and shardId instead of `placementState` to make this more explicit.

Also, I think having all of the logic for switch-over helps readability.